contracts/BonusRewards.sol

```diff
@@ -3,11 +3,340 @@
+        require(bonus.bonusTokenAddr == _token, "BonusRewards: wrong pool");
+        require(bonus.endTime + WEEK < block.timestamp, "BonusRewards: not ready");
+        balance = bonus.remBonus;
+        pools[_lpToken].bonuses[_poolBonusId].remBonus = 0;
+      }
+
+      IERC20(_token).transfer(owner(), balance);
+    }
+  }
+
+  function setResponders(address[] calldata _responders) external override onlyOwner {
+    responders = _responders;
+  }
+
+  function setPaused(bool _paused) external override {
+    require(_isAuthorized(msg.sender, responders), "BonusRewards: caller not responder");
+    paused = _paused;
+  }
+
+  function _updateUserWriteoffs(address _lpToken) private {
+    Bonus[] memory bonuses = pools[_lpToken].bonuses;
+    User storage user = users[_lpToken][msg.sender];
+    for (uint256 i = 0; i < bonuses.length; i++) {
+      // update writeoff to match current acc rewards per token
+      if (user.rewardsWriteoffs.length == i) {
+        user.rewardsWriteoffs.push(user.amount * bonuses[i].accRewardsPerToken / CAL_MULTIPLIER);
+      } else {
+        user.rewardsWriteoffs[i] = user.amount * bonuses[i].accRewardsPerToken / CAL_MULTIPLIER;
+      }
+    }
+  }
+
+  /// @notice tranfer upto what the contract has
+  function _safeTransfer(address _token, uint256 _amount) private returns (uint256 _transferred) {
+    IERC20 token = IERC20(_token);
+    uint256 balance = token.balanceOf(address(this));
+    if (balance > _amount) {
+      token.safeTransfer(msg.sender, _amount);
+      _transferred = _amount;
+    } else if (balance > 0) {
+      token.safeTransfer(msg.sender, balance);
+      _transferred = balance;
+    }
+  }
+
+  function _calRewardsForTime(Bonus memory _bonus, uint256 _lastUpdatedAt) internal view returns (uint256) {
+    if (_bonus.endTime <= _lastUpdatedAt) return 0;
+
+    uint256 calEndTime = block.timestamp > _bonus.endTime ? _bonus.endTime : block.timestamp;
+    uint256 calStartTime = _lastUpdatedAt > _bonus.startTime ? _lastUpdatedAt : _bonus.startTime;
+    uint256 timePassed = calEndTime - calStartTime;
+    return _bonus.weeklyRewards * CAL_MULTIPLIER * timePassed / WEEK;
+  }
+
+  function _claimRewards(address _lpToken, User memory _user) private {
+    // only claim if user has deposited before
+    if (_user.amount > 0) {
+      uint256 rewardsWriteoffsLen = _user.rewardsWriteoffs.length;
+      Bonus[] memory bonuses = pools[_lpToken].bonuses;
+      for (uint256 i = 0; i < bonuses.length; i++) {
+        uint256 rewardsWriteoff = rewardsWriteoffsLen <= i ? 0 : _user.rewardsWriteoffs[i];
+        uint256 bonusSinceLastUpdate = _user.amount * bonuses[i].accRewardsPerToken / CAL_MULTIPLIER - rewardsWriteoff;
+        if (bonusSinceLastUpdate > 0) {
+          uint256 transferred = _safeTransfer(bonuses[i].bonusTokenAddr, bonusSinceLastUpdate); // transfer bonus tokens to user
+          pools[_lpToken].bonuses[i].remBonus = bonuses[i].remBonus - transferred;
```

Comment:
It is possible for a malicious user to force a revert a here. If collectDust has been called while users still have some rewards outstanding. Then bonusSinceLastUpdate is a positive number. This is fine normally as balance of rewards is 0. But if a user send 1 wei of bonusToken we now revert.

Here is how to reproduce: 

Wait until after endTime. Then:

await bonusRewards..updatePool(lpToken.address);
await bonusRewards.collectDust(bonusToken.address, lpToken.address, 0);
await bonusToken.transfer(bonusRewards.address, 1);    
await bonusRewards.connect(userAAccount).claimRewards(lpToken.address);

Reply:
How can this be avoided?

Reply:
Good point. Tho the idea is that after a week of the ending of a bonus program, users have 7 days to claim the remaining rewards. After 7 days, if collectDust is called, then users will not be able (revert or transfer 0) to call `claimRewards` anymore.

Reply:
"After 7 days, if collectDust is called, then users will not be able (revert or transfer 0)". Revert is worse than transfer 0 because if there is a new bonus token added that will revert too. I don't think it is a big issue and is very much an edge case. 

In viewBonus you check if (bonus.remBonus > 0). You could do that here and it would stop the problem. if remBonus == 0 we can assume the bonus is deactivated. it should also save on gas

Reply:
good point. will update

Reply:
Update looks good